src/full.rs

```diff
@@ -0,0 +1,157 @@
+use crate::{Body, SizeHint};
+use bytes::{Buf, Bytes};
+use http::HeaderMap;
+use std::borrow::Cow;
+use std::convert::{Infallible, TryFrom};
+use std::marker::Unpin;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// A body that consists of a single chunk.
+#[derive(Clone, Copy, Debug)]
+pub struct Full<D> {
+    data: Option<D>,
+}
+
+impl<D> Full<D>
+where
+    D: Buf + Unpin,
+{
+    /// Create a new `Full`.
+    pub fn new(data: D) -> Self {
+        let data = if data.has_remaining() {
+            Some(data)
+        } else {
+            None
+        };
+        Full { data }
+    }
+}
+
+impl<D> Body for Full<D>
+where
+    D: Buf + Unpin,
+{
+    type Data = D;
+    type Error = Infallible;
+
+    fn poll_data(
+        mut self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<D, Self::Error>>> {
+        Poll::Ready(self.data.take().map(Ok))
+    }
+
+    fn poll_trailers(
+        self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<HeaderMap>, Self::Error>> {
+        Poll::Ready(Ok(None))
+    }
+
+    fn is_end_stream(&self) -> bool {
+        self.data.is_none()
+    }
+
+    fn size_hint(&self) -> SizeHint {
+        self.data
+            .as_ref()
+            .map(|data| {
+                u64::try_from(data.remaining())
+                    .map(SizeHint::with_exact)
+                    .unwrap_or_else(|_| {
+                        let mut hint = SizeHint::new();
+                        hint.set_lower(u64::MAX);
+                        hint
+                    })
+            })
+            .unwrap_or_else(|| SizeHint::with_exact(0))
+    }
+}
+
+impl<D> Default for Full<D>
+where
+    D: Buf + Unpin,
+{
+    /// Create an empty `Full`.
+    fn default() -> Self {
+        Full { data: None }
+    }
+}
+
+impl<D> From<Bytes> for Full<D>
+where
+    D: Buf + From<Bytes> + Unpin,
```

Comment:
I'm not used to seeing `impl`s that propagate a `From` like this. Is there any potential conflicts this could cause, if we wanted to provide concrete `From`s as well? For example, hyper's `From`s on `Body`? Not saying it's wrong, just checking before we merge :)

Reply:
I'm sorry but I still don't quite get what kind of code you are expecting a possibility of breakage.

That being said, among the currently known `Buf` implementations, the `impl`s seem to be only relevant to `Bytes`, `Box<Bytes>` (`From<Bytes> for Full<Box<Bytes>>`) and `VecDeque` (`From<Vec<u8>> for Full<VecDeque>`) and I'm not sure if the generics over `D` here is worth the complexity.

Maybe we should start conservatively and implement these `From`s for `Full<Bytes>` only for now?

Reply:
Yea, I'm happy starting conservative and adding as needed. Thanks!